stochastic fail with distributed tests

`testgather_all_tensors_distributed` fails stochastically.

At the same commit, the first attempt fails, while the second runs fine:

1. https://github.com/goodfire-ai/spd/actions/runs/18410304382/attempts/1

2. https://github.com/goodfire-ai/spd/actions/runs/18410304382/




## Error Message

```python
___________ TestGatherAllTensors.testgather_all_tensors_distributed ____________
[gw1] linux -- Python 3.12.11 /__w/spd/spd/.venv/bin/python3

self = <test_gather_all_tensors_distributed.TestGatherAllTensors object at 0x7f810b318860>

    def testgather_all_tensors_distributed(self):
        """Run distributed tests via mpirun in subprocess."""
        script_path = Path(__file__).resolve()
    
        # ports should be globally unique in tests to allow test parallelization
        # see discussion at: https://github.com/goodfire-ai/spd/pull/186
        env = {
            "MASTER_PORT": "29503",
            "OMP_NUM_THREADS": "1",
        }
    
        cmd = ["mpirun", "-np", "2", sys.executable, str(script_path)]
    
        result = subprocess.run(
            cmd, env={**os.environ, **env}, capture_output=True, text=True, timeout=120
        )
    
        if result.returncode != 0:
            print(f"STDOUT:\n{result.stdout}")
            print(f"STDERR:\n{result.stderr}")
>           raise RuntimeError(f"Distributed test failed with code {result.returncode}")
E           RuntimeError: Distributed test failed with code 134

tests/test_gather_all_tensors_distributed.py:233: RuntimeError
----------------------------- Captured stdout call -----------------------------
STDOUT:
[Gloo] Rank 0 is connected to 1 peer ranks. Expected number of connected peer ranks is : 1
[Gloo] Rank 1 is connected to 1 peer ranks. Expected number of connected peer ranks is : 1

Running 6 gather_all_tensors tests...

✓ Gather identical shapes test passed
✓ Gather scalar tensors test passed
✓ Gather multidimensional tensors test passed
✓ Gather empty tensor test passed
✓ Gather float tensor test passed
✓ Gather preserves autograd test passed

✓ All 6 distributed tests passed!


STDERR:
terminate called without an active exception
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpirun noticed that process rank 1 with PID 0 on node f2882a554fb7 exited on signal 6 (Aborted).
--------------------------------------------------------------------------
```



---

EDIT: another instance of a stochastic fail: https://github.com/goodfire-ai/spd/actions/runs/18412998079/attempts/1

--- a/tests/test_gather_all_tensors_distributed.py
+++ b/tests/test_gather_all_tensors_distributed.py
@@ -169,39 +169,38 @@ def run_all_tests():
     """Run all distributed tests when called directly with mpirun."""
     # Initialize distributed once for all tests
     init_distributed(backend="gloo")
-    rank = get_rank()
-    world_size = get_world_size()
+    try:
+        rank = get_rank()
+        world_size = get_world_size()
 
-    if world_size != 2:
+        assert world_size == 2, f"Tests require exactly 2 ranks, got {world_size}"
+
+        tests = [
+            ("Gather identical shapes", _test_gather_identical_shapes),
+            ("Gather scalar tensors", _test_gather_scalar_tensors),
+            ("Gather multidimensional tensors", _test_gather_multidimensional_tensors),
+            ("Gather empty tensor", _test_gather_empty_tensor),
+            ("Gather float tensor", _test_gather_float_tensor),
+            ("Gather preserves autograd", _test_gather_preserves_autograd),
+        ]
+
         if rank == 0:
-            print(f"✗ Tests require exactly 2 ranks, got {world_size}")
+            print(f"\nRunning {len(tests)} gather_all_tensors tests...\n")
+
+        for test_name, test_func in tests:
+            try:
+                test_func()
+            except Exception as e:
+                if rank == 0:
+                    print(f"✗ {test_name} failed: {e}")
+                raise
+            # Small barrier to ensure clean test separation
+            sync_across_processes()
+
+        if rank == 0:
+            print(f"\n✓ All {len(tests)} distributed tests passed!\n")
+    finally:
         cleanup_distributed()
-        sys.exit(1)
-
-    tests = [
-        ("Gather identical shapes", _test_gather_identical_shapes),
-        ("Gather scalar tensors", _test_gather_scalar_tensors),
-        ("Gather multidimensional tensors", _test_gather_multidimensional_tensors),
-        ("Gather empty tensor", _test_gather_empty_tensor),
-        ("Gather float tensor", _test_gather_float_tensor),
-        ("Gather preserves autograd", _test_gather_preserves_autograd),
-    ]
-
-    if rank == 0:
-        print(f"\nRunning {len(tests)} gather_all_tensors tests...\n")
-
-    for test_name, test_func in tests:
-        try:
-            test_func()
-        except Exception as e:
-            if rank == 0:
-                print(f"✗ {test_name} failed: {e}")
-            raise
-        # Small barrier to ensure clean test separation
-        sync_across_processes()
-
-    if rank == 0:
-        print(f"\n✓ All {len(tests)} distributed tests passed!\n")
 
 
 # ===== Pytest wrapper =====

